### PR TITLE
[Dispatcher] Collect autograd sequence numbers on PythonTLSSnapshot dispatch keys

### LINKED ISSUE #121758: Profiler traces for python subclasses are missing fwd-bwd links
## TL;DR

When you run the profiler with tensors that require grad (and run both the forward and backward passes), you should get links that connect the forward and backward kernels:

Screenshot 2024-03-12 at 2 11 12 PM

But when the tensor is a Nested Tensor (NT) (or, presumably any type of python tensor subclass?), the links don't appear

## Repro
```python
import torch
from torch.nested._internal.nested_tensor import ViewNestedFromBuffer, buffer_from_jagged

# set this to "False" and the test will stop failing
use_nt = True

with torch.profiler.profile() as prof:
    values = torch.randn(4, 6, requires_grad=True)
    offsets = torch.tensor([0,2,4])
    values = values * 2
    l = torch.nn.Linear(6,8)
    if use_nt:
        nt = ViewNestedFromBuffer.apply(values, offsets)
    else:
        nt = values

    nt = l(nt)
    if use_nt:
        val = buffer_from_jagged(nt)
    else:
        val = nt

    loss = val.sum()
    loss.backward()

found_seq_nr = False
for evt in prof.events():
    if "linear" in evt.name and "backward" not in evt.name:
        found_seq_nr = found_seq_nr or evt.sequence_nr != -1

assert found_seq_nr
```

## Reason

**How fwd-bwd links work**: We collect a "sequence number" from the autograd engine that is attached to every autograd-enabled forward kernel; the same sequence number is attached to the corresponding backward kernel. The profiler collects the sequence number for the forward and backward kernel, and then does some post-processing to connect them.

**Why we don't see fwd-bwd links for NT**: 

The record_function added in the dispatcher (which is responsible for adding the profiler events) does not always get called for every dispatch event.

https://github.com/pytorch/pytorch/blob/522d9729247d30cc4e96437023e66a516add4c04/aten/src/ATen/core/dispatch/Dispatcher.cpp#L511-L519

We'll only include the sequence number if we find that the current dispatch key that we're handling is an Autograd key.

*Debugging tip: build with DEBUG=1 and run with TORCH_SHOW_DISPATCH_TRACE=1 to see a list of all the dispatch calls.*

If we look at the `linear` call, we see:

```
 [call] op=[aten::linear], key=[PythonTLSSnapshot]
  [redispatchBoxed] op=[aten::linear], key=[AutogradNestedTensor]
   [redispatch] op=[aten::linear], key=[Python]
    [callBoxed] op=[aten::linear], key=[CPU]
     [call] op=[aten::t], key=[CPU]
      [call] op=[aten::transpose.int], key=[CPU]
       [call] op=[aten::as_strided], key=[CPU]
     [call] op=[aten::addmm], key=[CPU]
      [call] op=[aten::expand], key=[CPU]
       [call] op=[aten::as_strided], key=[CPU]
      [call] op=[aten::copy_], key=[CPU]
      [call] op=[aten::resolve_conj], key=[CPU]
      [call] op=[aten::resolve_conj], key=[CPU]
```

and, when we compare with a debugger, we see that the recordFunction is only called for `[call] op=[aten::linear], key=[PythonTLSSnapshot]` and for `[callBoxed] op=[aten::linear], key=[CPU]`. This checks out; when we see the profiler trace, we only see 2 `linear` calls, not the 4 shown by the dispatch trace debug shown above.

## Solution ideas

**1. Collect more dispatch events**

 i.e. also collect `[redispatch]` and `[redispatchBoxed]`. This is probably somewhat straightforward; now we'll collect all the events including the Autograd events that will have the sequence number attached. *However*, this will probably increase trace sizes globally, which is bad in itself and may have additional effects depending on complexity of downstream processing of traces.

**2. Relax condition for attaching sequence_nr**

Maybe do something like: "if the current dispatch key set contains autograd, then include sequence_nr". Then we'd include sequence_nr for the PythonTLSSnapshot event, even though it is not directly an Autograd dispatch key. *Problem*: we may introduce multiple events with the same sequence_nr; we'll have to check the fwd-bwd postprocessing to see how this will be handled, and in general it would be confusing.

**3. Special handling for PythonTLSSnapshot?**

We seem to be calling PythonTLSSnapshot in this case - maybe we can handle this dispatch key specially?

* redispatchBoxed vs. callBoxed: what's the difference? Can we callBoxed instead of redispatchBoxed?
* Can we assume that if PythonTLSSnapshot is called, then we'll never call Autograd directly internally? If so, can we put the sequence_nr here?
* AutogradNestedTensor handling - is it special?

### Alternatives

_No response_

### Additional context

_No response_

cc @robieta @chaekit @aaronenyeshi @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @dzhulgakov @ezyang @msaroufim @albanD

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123304

Fixes #121758

**TL;DR**: When profiling is turned on, the dispatcher will sometimes attach the autograd sequence number to the recorded profiler event. This PR expands the set of profiler events onto which we attach sequence numbers. Before, we'd only attach a sequence number if the current dispatch key was an Autograd dispatch key. Now, we attach a sequence number if the current dispatch key **set** contains Autograd.

**Context**:
The use case for this is torch.profiler for python subclasses.

Autograd attaches a "sequence number" to all ops that it encounters during the forward pass. Then, the corresponding sequence number can be associated with a backward kernel when backward is executed. This is used by the profiler to associate the forward ops to the backward ops; a forward op and a backward op with the same sequence number are "linked" in some post-processing step.

Prior to this PR, this profiler feature didn't work for python subclasses. The reason is that we don't collect profiler information for all the dispatches for a given kernel; we only dispatch the initial `call`, and not the subsequent `redispatch` invocations. Normally, an Autograd key (if we're running with autograd) is the highest dispatch key, so the initial `call` that we profile is an Autograd key, and we collect the sequence number. But when we're dealing with a python subclass, the first dispatch key is PythonTLSSnapshot, which eventually redispatches into Autograd. We don't record the Autograd sequence number in that case because we don't record redispatches.

To fix this, this PR collects a sequence number whenever the dispatch key **set** contains an Autograd key. That means we might sometimes collect multiple events with the same sequence number, or possibly attach sequence numbers when we won't actually use them? (e.g. maybe if the initial dispatch key handler removes Autograd for some reason). Although this might be a bit confusing for users looking directly at the sequence_nr directly, I think the main use case is for the profiler to create fwd-bwd links; and those should still be generated correctly in these cases.

Differential Revision: [D55724190](https://our.internmc.facebook.com/intern/diff/D55724190)

cc @ezyang @msaroufim @albanD

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55724190).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55724190).

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 1]
```diff
@@ -498,24 +498,44 @@ std::vector<OperatorName> Dispatcher::getRegistrationsForDispatchKey(c10::option
...
+  // The first entry into the dispatch is usually going to be a call(), but
+  // after that we usually redispatch().
+  // We only profile call() and callBoxed(), not redispatch.*()
+  // dispatcher calls, and when running with python subclasses, the top-level
+  // call is a PythonTLSSnapshot call, not an Autograd call. That means that
+  // we need special handling for subclasses if we want to collect the
+  // sequence number for operators on python subclasses.
+  //
+  // This is the special case: if we're profiling a PythonTLSSnapshot call
+  // and we have an autograd dispatch key in the dispatchKeySet, then we'll
```
Comment (albanD) [inline aten/src/ATen/core/dispatch/Dispatcher.cpp:520]: I'm curious why the logic here is not simplified to always get the seq_num unless autograd is in the exclude set?

Comment (davidberard98) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp:520]: @albanD that's a good point.

My goal was to only capture the sequence_nr once during the forward pass in profiler events. My guess was that this is a safe assumption for PythonTLSSnapshot, but that it might not be true in general (i.e. if we `call` PythonTLSSnapshot, then we won't later `call` again on or above Autograd).

But I guess maybe I should instead just always get the sequence_nr and then fix the profiler logic so that it handles duplicate sequence_nr in forward. Let me experiment to see if I can repro this case and try to fix it

Comment (albanD) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp:520]: From what I recall, the profiler already has to handle duplicates since the current logic can actually trigger multiple peek() of the same number when CompositeImplicit ops are involved.

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55724190).

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 2]
```diff
@@ -498,24 +498,38 @@ std::vector<OperatorName> Dispatcher::getRegistrationsForDispatchKey(c10::option
...
+  // This means that we might collect the same sequence nubmer two different
+  // events if they all occurred above Autograd and still had the Autograd
+  // dispatch key in the dispatch key set.
+  // However, this usually doesn't happen: normally the first call will
+  // go through the call() or callBoxed() path in the dispatcher, while
+  // subsequent redispatches go through redispatch() or redispatchBoxed().
+  // `call` has profiler instrumentation, whereas `redispatch` doesn't.
+  // So usually, we'll collect a sequence number on the first call() if the
+  // dispatch keys contain autograd, and not on subsequent redispatches.
+  bool dispatchHasAutograd = !(dispatchKeySet & autograd_dispatch_keyset).empty();
```
Comment (albanD) [inline aten/src/ATen/core/dispatch/Dispatcher.cpp:517]: btw my suggestion was to check the exclude set here.
The way autograd works is that the key is added to every single Tensor. So everything will always dispatch to autograd unless:
- Autograd key is in the exclude set
- Inference mode is running

Since inference mode implies no_grad mode (checked below), I think it is enough to check the tls exclude set here. That way you don't need to pass any argument in and can simplify the overall logic quite a bit.

Comment (davidberard98) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp:517]: ah I see, thanks! I figured we didn't need to check given we checked no_grad mode below, but your explanation makes it clear why checking the exclude set is better

Comment (davidberard98) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp:517]: @albanD - actually, I think the exclude set should already be handled:

from callBoxed, this gets the dispatch key set, which is eventually passed into the sequenceNumberForRunningRecordFunction function that I'm modifying in this PR:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/Dispatcher.h#L718

and getDispatchKeySetBoxed eventually calls computeDispatchKeySet:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L170

computeDispatchKeySet checks the exclude set:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L46

Based on this, do you think it's fine to leave it as is? (Is GradMode expected to show the exact same info as the include/exclude dispatch key sets? I checked and it wasn't obvious to me)

Comment (soulitzer) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp:517]: gradmode and include/exclude can diverge for example (if this is what you are asking) because when you redispatch in the autograd kernel, you first disable autograd via the exclude key. If you later have a callBoxed  in torch dispatch, gradmode can be still enabled, but the autograd key has now been excluded.

Comment (soulitzer) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp:517]: > actually, I think the exclude set should already be handle

I think the idea is that if checking only exclude set is sufficient to determine whether autograd kernel is reached, then we can avoid threading dispatch key at all these callsites and instead just check TLS inside sequenceNumberForRunningRecordFunction

However, if you create tensors within inference mode and then use them outside of inference mode, you will not reach the autograd kernel even though you are outside of no_grad, and autograd is not in TLS exclude set.

So, TLDR - the current appraoch is probably fine.

**Review by soulitzer — APPROVED** [review]
Thanks!
